bigquery_schema_generator/generate_schema.py

```diff
@@ -348,11 +355,20 @@ def merge_schema_entry(
         # upgraded to a REPEATED {primitive_type}, but currently 'bq load' does
         # not support that so we must also follow that rule.
         if old_mode != new_mode:
-            self.log_error(
-                f'Ignoring non-RECORD field with mismatched mode: '
-                f'old=({old_status},{full_old_name},{old_mode},{old_type}); '
-                f'new=({new_status},{full_new_name},{new_mode},{new_type})')
-            return None
+            # primitive-types are conservatively deduced NULLABLE. In case we
```

Comment:
Hi @bxparks this came from @bozzzzo's initial implementation.

I implemented your suggestion and then when writing a unit testcase to go from REQUIRED --> NULLABLE so I could ensure errors existed properly I could never trigger this. After a lot of tracing this is because when we deduce an existing schema we assume that status is hard since we can't change a type on line 286
```python
        # new 'soft' does not clobber old 'hard'
        if old_status == 'hard' and new_status == 'soft':
            return old_schema_entry
```


Given we assume all are hard and if we find a NULL field in a schema_entry we instantiate it with a status of `soft` we will never reach the condition to relax from REQUIRED --> NULLABLE.

Code for instantiating a NULL schema_entry https://github.com/bxparks/bigquery-schema-generator/blob/7d1b4ce892a174bc9dfaa6b5b3220691919f424f/bigquery_schema_generator/generate_schema.py#L427-L436


Should we move the logic about relaxing a field from REQUIRED --> NULLABLE or NULLABLE --> REQUIRED further up in this function or should we delete this status check and use it only when doing type conversions?

Reply:
I broke out the mode determination into it's own function and I call that within this soft/hard check and lower in the code where this comment initially referenced and it _seems_ to be working. Expect a commit with the change and updated testcase soon.

Reply:
This has now been implemented with tests and pushed in bb5745c8b55e7d4b51a191feab995bc0cb308c81